x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiUnifiedStreamingProcessor.java

```diff
@@ -0,0 +1,262 @@
+ *              ]
+ *          }
+ *
+ *         {
+ *              "id": "chatcmpl-123",
+ *              "object": "chat.completion",
+ *              "created": 1677652288,
+ *              "model": "gpt-3.5-turbo-0613",
+ *              "system_fingerprint": "fp_44709d6fcb",
+ *              "choices": [
+ *                  {
+ *                      "index": 1,
+ *                      "delta": {
+ *                          "content": "may I assist you today?",
+ *                      },
+ *                      "finish_reason": ""
+ *                  }
+ *              ]
+ *          }
+ *
+ *         {
+ *              "id": "chatcmpl-123",
+ *              "object": "chat.completion",
+ *              "created": 1677652288,
+ *              "model": "gpt-3.5-turbo-0613",
+ *              "system_fingerprint": "fp_44709d6fcb",
+ *              "choices": [
+ *                  {
+ *                      "index": 2,
+ *                      "delta": {},
+ *                      "finish_reason": "stop"
+ *                  }
+ *              ]
+ *          }
+ *
+ *          [DONE]
+ *     </code>
+ * </pre>
+ */
+public class OpenAiUnifiedStreamingProcessor extends DelegatingProcessor<Deque<ServerSentEvent>, ChunkedToXContent> {
+    private static final Logger log = LogManager.getLogger(OpenAiUnifiedStreamingProcessor.class);
+    private static final String FAILED_TO_FIND_FIELD_TEMPLATE = "Failed to find required field [%s] in OpenAI chat completions response";
+
+    private static final String CHOICES_FIELD = "choices";
+    private static final String DELTA_FIELD = "delta";
+    private static final String CONTENT_FIELD = "content";
+    private static final String DONE_MESSAGE = "[done]";
+    private static final String REFUSAL_FIELD = "refusal";
+    private static final String TOOL_CALLS_FIELD = "tool_calls";
+
+    @Override
+    protected void next(Deque<ServerSentEvent> item) throws Exception {
+        var parserConfig = XContentParserConfiguration.EMPTY.withDeprecationHandler(LoggingDeprecationHandler.INSTANCE);
+
+        var results = new ArrayDeque<StreamingUnifiedChatCompletionResults.Result>(item.size());
+        for (ServerSentEvent event : item) {
+            if (ServerSentEventField.DATA == event.name() && event.hasValue()) {
+                try {
+                    var delta = parse(parserConfig, event);
+                    delta.forEachRemaining(results::offer);
+                } catch (Exception e) {
+                    log.warn("Failed to parse event from inference provider: {}", event);
+                    throw e;
+                }
+            }
+        }
+
+        if (results.isEmpty()) {
+            upstream().request(1);
+        } else {
+            downstream().onNext(new StreamingUnifiedChatCompletionResults.Results(results));
+        }
+    }
+
+    private Iterator<StreamingUnifiedChatCompletionResults.Result> parse(XContentParserConfiguration parserConfig, ServerSentEvent event)
```

Comment:
This code is quite similar to `OpenAiStreamingProcessor` especially the parsing logic. Could we replace `OpenAiStreamingProcessor` with this implementation and throw away the tool calls if not required.

One option is a pass in a function that creates the result (`StreamingChatCompletionResults.Result` or `StreamingUnifiedChatCompletionResults.Result`) given `content, refusal, toolCalls` as the input